File: externals/Async-CTP/samples/VB/Windows/Phone/Concert/Finder/Concert/Finder/ViewModel/MainPageViewModel.vb
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows
Imports Microsoft.Phone.Reactive
Imports Concert_Finder.ConcertFinder.Model

Namespace ConcertFinder.ViewModel

    ''' <summary>
    ''' The view model for the MainPage.
    ''' </summary>
    Public Class MainPageViewModel
        Implements INotifyPropertyChanged

        ''' <summary>
        ''' The list of search results.
        ''' </summary>
        Public Property SearchResults As ObservableCollection(Of ISearchable)

        Public ReadOnly Property IsListEmpty As Boolean
            Get
                Return Not _SearchPending AndAlso Not SearchInProgress AndAlso SearchResults.Count = 0
            End Get
        End Property

        ''' <summary>
        ''' Gets or sets whether a search is in progress.
        ''' </summary>
        Public Property SearchInProgress As Boolean
            Get
                Return _SearchInProgress
            End Get
            Set(value As Boolean)
                _SearchInProgress = value
                _SearchPending = False
                NotifyPropertyChanged("SearchInProgress")
                NotifyPropertyChanged("IsListEmpty")
            End Set
        End Property
        Private _SearchInProgress As Boolean = False
        Private _SearchPending As Boolean = True

        ''' <summary>
        ''' The default constructor.
        ''' </summary>
        Public Sub New()
            SearchResults = New ObservableCollection(Of ISearchable)()
        End Sub

        ''' <summary>
        ''' Perform a search based on the user's query.
        ''' </summary>
        ''' <param name="query">The search query.</param>
        Public Sub Search(query As String)
            Observable.FromEvent(Of SearchResultsAvailableEventArgs)(
                Sub(ev) AddHandler App.SearchResultsAvailable, ev,
                Sub(ev) RemoveHandler App.SearchResultsAvailable, ev).
            [Take](1).Subscribe(
                Sub(e)
                    Deployment.Current.Dispatcher.BeginInvoke(
                        Sub()
                            SearchResults.Clear()

                            For Each SearchResult In e.EventArgs.SearchResults
                                SearchResults.Add(SearchResult)
                            Next

                            SearchInProgress = False
                        End Sub)
                End Sub
            )

            SearchInProgress = True

            App.Search(query)
        End Sub

        ''' <summary>
        ''' The PropertyChanged event.
        ''' </summary>
        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        ''' <summary>
        ''' Notify listeners of property changes.
        ''' </summary>
        ''' <param name="propertyName">The name of the changed property.</param>
        Private Sub NotifyPropertyChanged(propertyName As String)
            RaiseEvent PropertyChanged(Nothing, New PropertyChangedEventArgs(propertyName))
        End Sub

    End Class

End Namespace

File: externals/Async-CTP/samples/CS/Windows/Phone/Concert/Finder/Concert/Finder/ViewModel/MainPageViewModel.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Windows;
using ConcertFinder.Model;
using Microsoft.Phone.Reactive;

namespace ConcertFinder.ViewModel
{
    /// <summary>
    /// The view model for the MainPage.
    /// </summary>
    public class MainPageViewModel : INotifyPropertyChanged
    {
        /// <summary>
        /// The list of search results.
        /// </summary>
        public ObservableCollection<ISearchable> SearchResults { get; set; }
        public bool IsListEmpty { get { return !_SearchPending && !SearchInProgress && SearchResults.Count == 0; } }

        /// <summary>
        /// Gets or sets whether a search is in progress.
        /// </summary>
        public bool SearchInProgress
        {
            get { return _SearchInProgress; }
            set
            {
                _SearchInProgress = value;
                _SearchPending = false;
                NotifyPropertyChanged("SearchInProgress");
                NotifyPropertyChanged("IsListEmpty");
            }
        }
        private bool _SearchInProgress = false;
        private bool _SearchPending = true;

        /// <summary>
        /// The default constructor.
        /// </summary>
        public MainPageViewModel()
        {
            SearchResults = new ObservableCollection<ISearchable>();
        }

        /// <summary>
        /// Perform a search based on the user's query.
        /// </summary>
        /// <param name="query">The search query.</param>
        public void Search(string query)
        {
            Observable.FromEvent<SearchResultsAvailableEventArgs>(
                ev => App.SearchResultsAvailable += ev,
                ev => App.SearchResultsAvailable -= ev).
                Take(1).
                Subscribe(
                (e) =>
                {
                    Deployment.Current.Dispatcher.BeginInvoke(() =>
                    {
                        SearchResults.Clear();

                        foreach (var SearchResult in e.EventArgs.SearchResults)
                        {
                            SearchResults.Add(SearchResult);
                        }

                        SearchInProgress = false;
                    });
                });

            SearchInProgress = true;

            App.Search(query);
        }

        #region INotifyPropertyChanged

        /// <summary>
        /// The PropertyChanged event.
        /// </summary>
        public event PropertyChangedEventHandler PropertyChanged;

        /// <summary>
        /// Notify listeners of property changes.
        /// </summary>
        /// <param name="propertyName">The name of the changed property.</param>
        private void NotifyPropertyChanged(string propertyName)
        {
            var handler = PropertyChanged;

            if (handler != null)
            {
                handler(null, new PropertyChangedEventArgs(propertyName));
            }
        }

        #endregion
    }
}
